File: java/src/uk/ac/ebi/webservices/jdispatcher/AbstractWsClient.java
/* $Id$
 * ======================================================================
 * Abstract jDispatcher web services Java client.
 * ====================================================================== */
package uk.ac.ebi.webservices.jdispatcher;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.lang.reflect.InvocationTargetException;
import java.rmi.RemoteException;
import javax.xml.rpc.ServiceException;
import org.apache.commons.cli.*;

/** Abstract class defining common methods to all jDispatcher SOAP web 
 * service clients.
 * 
 * See:
 * <a href="http://www.ebi.ac.uk/Tools/webservices/tutorials/java">http://www.ebi.ac.uk/Tools/webservices/tutorials/java</a>
 */
public abstract class AbstractWsClient {
	/** Output level. Controlled by the --verbose and --quiet options. */
	protected int outputLevel = 1;
	/** Debug level. Controlled by the --debugLevel option. */
	private int debugLevel = 0;
	/** Maximum interval between polling events (ms) */
	private int maxCheckInterval = 60000;
	/** Generic options message */
	private static final String genericOptsStr = "[General]\n"
	+ "\n"
	+ "      --params         :      : list tool parameters\n"
	+ "      --paramDetail    : str  : information about a parameter\n"
	+ "      --email          : str  : e-mail address\n"
	+ "      --title          : str  : title for the job\n"
	+ "      --async          :      : perform an asynchronous submission\n"
	+ "      --jobid          : str  : job identifier\n"
	+ "      --status         :      : get status of a job\n"
	+ "      --resultTypes    :      : get list of result formats for a job\n"
	+ "      --polljob        :      : get results for a job\n"
	+ "      --outfile        : str  : name of the file results should be written to\n"
	+ "                                (default is based on the jobid; \"-\" for STDOUT)\n"
	+ "      --outformat      : str  : output format, see --resultTypes\n"
	+ "      --help           :      : prints this help text\n"
	+ "      --quiet          :      : decrease output\n"
	+ "      --verbose        :      : increase output\n"
	+ "      --debugLevel     : int  : set debug output level\n"
	+ "\n"
	+ "Synchronous job:\n"
	+ "\n"
	+ "  The results/errors are returned as soon as the job is finished.\n"
	+ "  Usage: java -jar <jarFile> --email <your@email> [options...] seqFile\n"
	+ "  Returns: results as an attachment\n"
	+ "\n"
	+ "Asynchronous job:\n"
	+ "\n"
	+ "  Use this if you want to retrieve the results at a later time. The results \n"
	+ "  are stored for up to 24 hours. \n"
	+ "  Usage: java -jar <jarFile> --async --email <your@email> [options...] seqFile\n"
	+ "  Returns: jobid\n"
	+ "\n"
	+ "  Use the jobid to query for the status of the job. If the job is finished, \n"
	+ "  it also returns the results/errors.\n"
	+ "  Usage: java -jar <jarFile> --polljob --jobid <jobId> [--outfile string]\n"
	+ "  Returns: string indicating the status of the job and if applicable, results \n"
	+ "  as an attachment.\n";

	/** Add genetic options to command-line parser. */
	protected static void addGenericOptions(Options options) {
		options.addOption("help", "help", false, "help on using this client");
		options.addOption("async", "async", false, "perform an asynchronous job");
		options.addOption("polljob", "polljob", false, "poll for the status of an asynchronous job and get the results");
		options.addOption("status", "status", false, "poll for the status of an asynchronous job");        
		options.addOption("email", "email", true, "Your email address");
		options.addOption("jobid", "jobid", true, "Job identifier of an asynchronous job");
		options.addOption("stdout", "stdout", false, "print to standard output");
		options.addOption("outfile", "outfile", true, "file name to save the results");
		options.addOption("outformat", "outformat", true, "Output format (txt or xml)");
		options.addOption("quiet", "quiet", false, "Decrease output messages");
		options.addOption("verbose", "verbose", false, "Increase output messages");
		options.addOption("params", "params", false, "List parameters");
		options.addOption("paramDetail", "paramDetail", true, "List parameter information");
		options.addOption("resultTypes", "resultTypes", false, "List result types for job");
		options.addOption("debugLevel", "debugLevel", true, "Debug output");
	}

	/** Print the generic options usage message to STDOUT.
	 */
	protected static void printGenericOptsUsage()  {
		System.out.println(genericOptsStr);
	}
	
	/** Set debug level. 
	 * 
	 * @param level Debug level.
	 */
	public void setDebugLevel(int level) {
		printDebugMessage("setDebugLevel", "Begin " + level, 1);
		if(level > -1) {
			debugLevel = level;
		}
		printDebugMessage("setDebugLevel", "End", 1);
	}

	/** Get current debug level. 
	 * 
	 * @return Debug level.
	 */
	public int getDebugLevel() {
		printDebugMessage("getDebugLevel", new Integer(debugLevel).toString(), 1);
		return debugLevel;
	}

	/** Output debug message at specified level
	 * 
	 * @param methodName Name of the method to appear in the message
	 * @param message The message
	 * @param level Level at which to output message
	 */
	protected void printDebugMessage(String methodName, String message, int level) {
		if(level <= debugLevel) {
			System.err.println("[" + methodName + "()] " + message);
		}
	}
	
	/** Generate a string containing the values of the fields with "get" methods. 
	 * 
	 * @param obj Object the get values from
	 * @return String containing values and method names.
	 */
	@SuppressWarnings("unchecked")
	protected String objectFieldsToString(Object obj) {
		printDebugMessage("ObjectFieldsToString", "Begin", 31);
		StringBuilder strBuilder = new StringBuilder();
		try {
			Class objType = obj.getClass();
			printDebugMessage("ObjectFieldsToString", "objType: " + objType, 32);
			java.lang.reflect.Method[] methods = objType.getMethods();
			for(int i = 0; i < methods.length; i++) {
				String methodName = methods[i].getName();
				if(methodName.startsWith("get") && methods[i].getParameterTypes().length == 0 &&
						!methodName.equals("getClass") && !methodName.equals("getTypeDesc")) {
					printDebugMessage("ObjectFieldsToString", "invoke(): " + methodName, 32);
					Object tmpObj = methods[i].invoke(obj, new Object[0]);
					// Handle any string lists (e.g. database)
					if(tmpObj instanceof String[]) {
						String[] tmpList = (String[])tmpObj;
						strBuilder.append(methodName + ":\n");
						for(int j = 0; j < tmpList.length; j++) {
							strBuilder.append("\t" + tmpList[j] + "\n");
						}
					}
					// Otherwise just use implicit toString();
					else {
						strBuilder.append(methodName + ": " + tmpObj + "\n");
					}
				}
			}
		}
		catch(SecurityException e) {
			System.err.println(e.getMessage());
		}
		catch (IllegalArgumentException e) {
			System.err.println(e.getMessage());
		}
		catch (IllegalAccessException e) {
			System.err.println(e.getMessage());
		}
		catch (InvocationTargetException e) {
			System.err.println(e.getMessage());
		}
		printDebugMessage("ObjectFieldsToString", "End", 31);
		return strBuilder.toString();
	}

	
	/** Set the output level. 
	 * 
	 * @param level Output level
	 */
	public void setOutputLevel(int level) {
		printDebugMessage("setOutputLevel", "Begin " + level, 1);
		if(level > -1) {
			this.outputLevel = level;
		}
		printDebugMessage("setOutputLevel", "End", 1);
	}

	/** Get the current output level.
	 * 
	 * @return Output level.
	 */
	public int getOutputLevel() {
		printDebugMessage("getOutputLevel", new Integer(this.outputLevel).toString(), 1);
		return this.outputLevel;
	}
	
	/** Set the maximum interval between polling events.
	 * 
	 * @param checkInterval Maximum interval in milliseconds
	 */
	public void setMaxCheckInterval(int checkInterval) {
		printDebugMessage("setMaxCheckInterval", "Begin " + checkInterval, 1);
		if(checkInterval > 1000) {
			this.maxCheckInterval = checkInterval;
		}
		printDebugMessage("setMaxCheckInterval", "End", 1);
	}
	
	/** Get the maximum interval between polling events.
	 * 
	 * @return Maximum interval in milliseconds
	 */
	public int getMaxCheckInterval() {
		printDebugMessage("getMaxCheckInterval", new Integer(this.maxCheckInterval).toString(), 1);
		return this.maxCheckInterval;
	}

	/** Print a progress message.
	 * 
	 * @param msg The message to print.
	 * @param level The output level at or above which this message should be displayed.
	 */
	protected void printProgressMessage(String msg, int level) {
		if(outputLevel >= level) {
			System.err.println(msg);
		}
	}

	/** Read the contents of a file into a byte array.
	 * 
	 * @param file the file to read
	 * @return the contents of the file in a byte array
	 * @throws IOException if all contents not read
	 */
	public byte[] readFile(File file) throws IOException, FileNotFoundException {
		printDebugMessage("readFile", "Begin", 1);
		printDebugMessage("readFile", "file: " + file.getPath() + File.pathSeparator + file.getName(), 2);
		if(!file.exists()) {
			throw new FileNotFoundException(file.getName());
		}
		InputStream is = new FileInputStream(file);
		byte[] bytes = readStream(is);
		is.close();
		printDebugMessage("readFile", "End", 1);
		return bytes;
	}

	/** Read the contents of an input stream into a string.
	 * 
	 * @param inStream the input steam to read
	 * @return the contents of the stream in a string
	 * @throws IOException if all contents not read
	 */
	public byte[] readStream(InputStream inStream) throws IOException {
		printDebugMessage("readStream", "Begin", 1);
		printDebugMessage("readStream", "End", 1);
		long length = inStream.available();
		byte[] bytes = new byte[(int)length];
		int offset = 0;
		int numRead = 0;
		while(offset < bytes.length &&
				(numRead=inStream.read(bytes,offset,bytes.length-offset)) >= 0 ) {
			offset += numRead;
		}
		if (offset < bytes.length) {
			throw new IOException("...");
		}
		printDebugMessage("readStream", "read " + bytes.length + " bytes", 2);
		return bytes;
	}

	/** Create data input structure from the option value.
	 * 
	 * The option can be either an entry identifier in the 
	 * format dbname:id or a filename. If a filename is used
	 * the contents of the file will be used as the input data. 
	 * This method makes no attempt to parse the input file to 
	 * handle it as individual sequences.
	 * 
	 * @param fileOptionStr Filename or entry identifier.
	 * @return Data to use as input as a byte array.
	 * @throws IOException
	 */
	public byte[] loadData(String fileOptionStr) throws IOException {
		printDebugMessage("loadData", "Begin", 1);
		printDebugMessage("loadData", "fileOptionStr: " + fileOptionStr, 2);
		byte[] retVal = null;
		if(fileOptionStr != null) {
			if(fileOptionStr.equals("-")) { // STDIN
				retVal = readStream(System.in);
			}
			else if(new File(fileOptionStr).exists()) { // File
				retVal = readFile(new File(fileOptionStr));
			} else { // Entry Id
				retVal = fileOptionStr.getBytes();
			}
		}
		printDebugMessage("loadData", "End", 1);
		return retVal;
	}

	/** Write a string to a file
	 * 
	 * @param file the file to create/write to
	 * @param data the string to write
	 * @return an integer value indicating success/failure
	 * @throws IOException if there is a problem with the file operations
	 */
	public int writeFile(File file, String data) throws IOException {
		printDebugMessage("writeFile", "Begin", 1);
		printDebugMessage("writeFile", "file: " + file.getName(), 2);
		printDebugMessage("writeFile", "data: " + data.length() + " characters", 2);
		OutputStream os = new FileOutputStream(file);
		PrintStream p = new PrintStream( os );
		p.println (data);
		p.close();
		printDebugMessage("writeFile", "End", 1);
		return 0;
	}
	
	/** Write bytes to a file
	 * 
	 * @param file the file to create/write to
	 * @param data the bytes to write
	 * @return an integer value indicating success/failure
	 * @throws IOException if there is a problem with the file operations
	 */
	public int writeFile(File file, byte[] data) throws IOException {
		printDebugMessage("writeFile", "Begin", 1);
		printDebugMessage("writeFile", "file: " + file.getName(), 2);
		printDebugMessage("writeFile", "data: " + data.length + " bytes", 2);
		OutputStream os = new FileOutputStream(file);
		os.write (data);
		os.close();
		printDebugMessage("writeFile", "End", 1);
		return 0;
	}
	
	/** Get an instance of the service proxy to use with other methods.
	 * 
	 * @throws ServiceException
	 */
	abstract protected void srvProxyConnect() throws ServiceException;

	/** Get list of tool parameter names.
	 * 
	 * @return String array containing list of parameter names
	 * @throws ServiceException
	 * @throws RemoteException
	 */
	abstract public String[] getParams() throws ServiceException, RemoteException;

	/** Print list of parameter names for tool.
	 * 
	 * @throws RemoteException
	 * @throws ServiceException
	 */
	protected void printParams() throws RemoteException, ServiceException {
		printDebugMessage("printParams", "Begin", 1);
		String[] paramList = getParams();
		for(int i = 0; i < paramList.length; i++) {
			System.out.println(paramList[i]);
		}
		printDebugMessage("printParams", "End", 1);
	}

	/** Print information about a tool parameter
	 * 
	 * @param paramName Name of the tool parameter to get information for.
	 * @throws RemoteException
	 * @throws ServiceException
	 */
	abstract protected void printParamDetail(String paramName) throws RemoteException, ServiceException;

	/** Get the status of a submitted job given its job identifier.
	 * 
	 * @param jobid The job identifier
	 * @return Job status as a string.
	 * @throws IOException
	 * @throws ServiceException
	 */
	abstract public String checkStatus(String jobid) throws IOException, ServiceException;

	/** Poll the job status until the job completes.
	 * 
	 * @param jobid The job identifier.
	 * @throws ServiceException
	 */
	public void clientPoll(String jobId) throws ServiceException {
		printDebugMessage("clientPoll", "Begin", 1);
		printDebugMessage("clientPoll", "jobId: " + jobId, 2);
		int checkInterval = 1000;
		String status = "PENDING";
		// Check status and wait if not finished
		while(status.equals("RUNNING") || status.equals("PENDING")) {
			try {
				status = this.checkStatus(jobId);
				printProgressMessage(status, 1);
				if(status.equals("RUNNING") || status.equals("PENDING")) {
					// Wait before polling again.
					printDebugMessage("clientPoll", "checkInterval: " + checkInterval, 2);
					Thread.sleep(checkInterval);
					checkInterval *= 2;
					if(checkInterval > this.maxCheckInterval) checkInterval = this.maxCheckInterval;
				}
			}
			catch(InterruptedException ex) {
				// Ignore
			}
			catch(IOException ex) {
				// Report and continue
				System.err.println("Warnning: " + ex.getMessage());
			}
		}
		printDebugMessage("clientPoll", "End", 1);
	}

	/** Print details of the available result types for a job.
	 * 
	 * @param jobId Job identifier to check for result types.
	 * @throws ServiceException
	 * @throws RemoteException
	 */
	abstract protected void printResultTypes(String jobId) throws ServiceException, RemoteException;

	/** Get the results for a job and save them to files.
	 * 
	 * @param jobid The job identifier.
	 * @param outfile The base name of the file to save the results to. If
	 * null the jobid will be used.
	 * @param outformat The name of the data format to save, e.g. toolraw 
	 * or toolxml. If null all available data formats will be saved.
	 * @return Array of filenames
	 * @throws IOException
	 * @throws javax.xml.rpc.ServiceException
	 */
	abstract public String[] getResults(String jobid, String outfile, String outformat) throws IOException, ServiceException;
}

File: csharp/EBIWS2_CSharp/EbiWS/AbstractWsClient.cs
/* $Id: AbstractWsClient.cs 1039 2009-06-06 20:36:34Z hpm $
 * ======================================================================
 * Common structure and methods for JDispatcher SOAP clients.
 * 
 * See:
 * http://www.ebi.ac.uk/Tools/webservices/
 * http://www.ebi.ac.uk/Tools/webservices/tutorials/csharp
 * ====================================================================== */
using System;
using System.IO;
using System.Reflection;
using System.Text;

namespace EbiWS
{
	/// <summary>
	/// Abstract definition of a client to the EMBL-EBI tool Web Services.
	/// </summary>
	public abstract class AbstractWsClient
	{
		/// <value>
		/// Level of output produced. Used to implment --quiet and --verbose.
		/// </value>
		public int OutputLevel {
			get{return outputLevel;}
			set{
				if(value > -1) outputLevel = value;
			}
		}
		private int outputLevel = 1;
		/// <value>
		/// Level of debug output (default off).
		/// </value>
		public int DebugLevel {
			get{return debugLevel;}
			set{
				if(value > -1) debugLevel = value;
			}
		}
		private int debugLevel = 0;
		/// <value>
		///  Maximum interval between status checks when polling a submited job.
		/// </value>
		public int MaxCheckInterval {
			get{return maxCheckInterval;}
			set{
				if(value > 5000) maxCheckInterval = value;
			}
		}
		private int maxCheckInterval = 60000;
		/// <value>
		/// Specified endpoint for the SOAP service. If null the default 
		/// endpoint specified in the WSDL (and thus in the generated 
		/// stubs) is used.
		/// </value>
		public string ServiceEndPoint {
			get{return serviceEndPoint;}
			set{serviceEndPoint = value;}
		}
		private string serviceEndPoint = null;
		/// <value>
		/// Parameter name to be used to get parameter details.
		/// </value>
		public string ParamName {
			get{return paramName;}
			set{paramName = value;}
		}
		private string paramName = null;
		/// <value>
		/// Output file name base
		/// </value>
		public string OutFile {
			get{return outFile;}
			set{outFile = value;}
		}
		private string outFile = null;
		/// <value>
		/// Output result format
		/// </value>
		public string OutFormat {
			get{return outFormat;}
			set{outFormat = value;}
		}
		private string outFormat = null;
		/// <value>
		/// User e-mail address for job submissions.
		/// </value>
		public string Email {
			get{return email;}
			set{email = value;}
		}
		private string email = null;
		/// <value>
		/// Title for job.
		/// </value>
		public string JobTitle {
			get{return jobTitle;}
			set{jobTitle = value;}
		}
		private string jobTitle = null;
		/// <value>
		/// Job Id for getting status or results
		/// </value>
		public string JobId {
			get{return jobId;}
			set{jobId = value;}
		}
		private string jobId = null;
		/// <value>
		/// Submission mode: async or sync
		/// </value>
		public bool Async {
			get{return async;}
			set{async = value;}
		}
		private bool async = false;
		/// <value>
		/// Action to perform
		/// </value>
		public string Action {
			get{return action;}
			set{action = value;}
		}
		private string action = "unknown";
		/// <summary>
		/// Usage message for generic options.
		/// </summary>
		private const string genericOptsStr = @"[General]

      --params         :      : list tool parameters
      --paramDetail    : str  : information about a parameter
      --email          : str  : e-mail address
      --title          : str  : title for the job
      --async          :      : perform an asynchronous submission
      --jobid          : str  : job identifier
      --status         :      : get status of a job
      --resultTypes    :      : get list of result formats for a job
      --polljob        :      : get results for a job
      --outfile        : str  : name of the file results should be written to
                                (default is based on the jobid; ""-"" for STDOUT)
      --outformat      : str  : output format, see --resultTypes
      --help           :      : prints this help text
      --quiet          :      : decrease output
      --verbose        :      : increase output
      --debugLevel     : int  : set debug output level

Synchronous job:

  The results/errors are returned as soon as the job is finished.
  Usage: tool.exe --email <your@email> [options...] seqFile
  Returns: results as an attachment

Asynchronous job:

  Use this if you want to retrieve the results at a later time. The results
  are stored for up to 24 hours.
  Usage: tool.exe --async --email <your@email> [options...] seqFile
  Returns: jobid

  Use the jobid to query for the status of the job. If the job is finished,
  it also returns the results/errors.
  Usage: tool.exe --polljob --jobid <jobId> [--outfile string]
  Returns: string indicating the status of the job and if applicable, results
  as an attachment.
";
		
		/// <summary>
		/// Default constructor.
		/// </summary>
		public AbstractWsClient()
		{
			OutputLevel = 1; // Normal output
			DebugLevel = 0; // Debug output off.
			MaxCheckInterval = 60000; // 1 min between checks
			OutFile = null;
			OutFormat = null;
			Email = null;
			JobTitle = "My Sequence";
			JobId = null;
			Async = false;
			Action = "UNKNOWN";
		}
		
		/// <summary>
		/// Print the generic options usage message to STDOUT.
		/// </summary>
		protected void PrintGenericOptsUsage()  {
			Console.WriteLine(genericOptsStr);
		}
		
		/// <summary>
		/// Print a debug message at the specified level.
		/// </summary>
		/// <param name="methodName">Method name to use in output.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="message">Message to output.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="level">Debug level at which to output.
		/// A <see cref="System.Int32"/>
		/// </param>
		protected void PrintDebugMessage(string methodName, string message, int level) {
			if(level <= DebugLevel) Console.Error.WriteLine("[{0}()] {1}", methodName, message);
		}

		/// <summary>
		/// Construct a string of the values of an object, both fields and properties.
		/// </summary>
		/// <param name="obj">
		/// Object to get values from. A <see cref="System.Object"/>
		/// </param>
		/// <returns>
		/// Name and the contents of values as a string. A <see cref="System.String"/>
		/// </returns>
		protected string ObjectValueToString(Object obj)
		{
			PrintDebugMessage("ObjectValueToString", "Begin", 31);
			StringBuilder strBuilder = new StringBuilder();
			strBuilder.Append(ObjectFieldsToString(obj));
			strBuilder.Append(ObjectPropertiesToString(obj));
			PrintDebugMessage("ObjectValueToString", "End", 31);
			return strBuilder.ToString();
		}

		/// <summary>
		/// Construct a string of the fields of an object.
		/// </summary>
		/// <param name="obj">
		/// Object to get fields from. A <see cref="System.Object"/>
		/// </param>
		/// <returns>
		/// Name and the contents of fields as a string. A <see cref="System.String"/>
		/// </returns>
		protected string ObjectFieldsToString(Object obj) {
			PrintDebugMessage("ObjectFieldsToString", "Begin", 32);
			StringBuilder strBuilder = new StringBuilder();
			Type objType = obj.GetType();
			PrintDebugMessage("ObjectFieldsToString", "objType: " + objType, 33);
			foreach(FieldInfo info in objType.GetFields()) {
				PrintDebugMessage("ObjectFieldsToString", "info: " + info.Name, 33);
				if (info.FieldType.IsArray)
				{
					strBuilder.Append(info.Name + ":\n");
					foreach(Object subObj in (Object[])info.GetValue(obj)) {
						strBuilder.Append("\t" + subObj);
					}
				}
				else {
					strBuilder.Append(info.Name + ": " + info.GetValue(obj) + "\n");
				}
			}
			PrintDebugMessage("ObjectFieldsToString", "End", 32);
			return strBuilder.ToString();
		}
		
		/// <summary>
		/// Construct a string of the properties of an object.
		/// </summary>
		/// <param name="obj">
		/// Object to get properties from. A <see cref="System.Object"/>
		/// </param>
		/// <returns>
		/// Name and the contents of properties as a string. A <see cref="System.String"/>
		/// </returns>
		protected string ObjectPropertiesToString(Object obj)
		{
			PrintDebugMessage("ObjectPropertiesToString", "Begin", 31);
			StringBuilder strBuilder = new StringBuilder();
			Type objType = obj.GetType();
			PrintDebugMessage("ObjectPropertiesToString", "objType: " + objType, 32);
			foreach (PropertyInfo info in objType.GetProperties())
			{
				PrintDebugMessage("ObjectPropertiesToString", "info: " + info.Name, 32);
				if (info.PropertyType.IsArray)
				{
					strBuilder.Append(info.Name + ":\n");
					foreach (Object subObj in (Object[])info.GetValue(obj, null))
					{
						strBuilder.Append("\t" + subObj);
					}
				}
				else
				{
					strBuilder.Append(info.Name + ": " + info.GetValue(obj, null) + "\n");
				}
			}
			PrintDebugMessage("ObjectPropertiesToString", "End", 31);
			return strBuilder.ToString();
		}
		
		/// <summary>
		/// Print a progress message, at the specified output level.
		/// </summary>
		/// <param name="msg">Message to print.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="level">Output level at which to print the message.
		/// A <see cref="System.Int32"/>
		/// </param>
		protected void PrintProgressMessage(String msg, int level) {
			if(OutputLevel >= level) Console.Error.WriteLine(msg);
		}
		
		/// <summary>
		/// Read data from a text file into a string.
		/// </summary>
		/// <param name="fileName">Name of the file to read data from.
		/// A <see cref="System.String"/>
		/// </param>
		/// <returns>Contents of the file as a string.
		/// A <see cref="System.String"/>
		/// </returns>
		protected string ReadTextFile(string fileName) {
			PrintDebugMessage("ReadTextFile", "Begin", 1);
			PrintDebugMessage("ReadTextFile", "fileName: " + fileName, 2);
			string retVal = "";
			// Read from STDIN
			if(fileName == "-") retVal = Console.In.ReadToEnd();
			// Read from file
			else retVal = File.ReadAllText(fileName);
			PrintDebugMessage("ReadTextFile", "read " + retVal.Length + " characters", 1);
			PrintDebugMessage("ReadTextFile", "End", 1);
			return retVal;
		}
		
		/// <summary>
		/// Read a file into a byte array.
		/// </summary>
		/// <param name="fileName">Filename to read data from.</param>
		/// <returns>Data read from file as a byte array.</returns>
		protected byte[] ReadFile(string fileName) {
			PrintDebugMessage("ReadFile", "Begin", 1);
			PrintDebugMessage("ReadFile", "fileName: " + fileName, 1);
			byte[] retVal = null;
			if(fileName == "-") { // Read from STDIN
				Stream s = Console.OpenStandardInput();
				BinaryReader sr = new BinaryReader(s);
				retVal = sr.ReadBytes((int)s.Length);
				// Do not close since this is STDIN.
			}
			else { // Read from file
				retVal = File.ReadAllBytes(fileName);
			}
			PrintDebugMessage("ReadFile", "read " + retVal.Length + " bytes", 1);
			PrintDebugMessage("ReadFile", "End", 1);
			return retVal;
		}
		
		/// <summary>
		/// Load text data to be submitted to the tool.
		/// </summary>
		/// <param name="fileOptionStr">Name of file to read data from, a file name of "-" reads data from standard input (STDIN), or raw data to be sent to tool.</param>
		/// <returns>Data as a string.</returns>
		protected string LoadData(string fileOptionStr) {
			PrintDebugMessage("LoadData", "Begin", 1);
			PrintDebugMessage("LoadData", "fileOptionStr: " + fileOptionStr, 2);
			string retVal = null;
			if(fileOptionStr != null) {
				if(fileOptionStr == "-") { // STDIN
					retVal = ReadTextFile(fileOptionStr);
				}
				else if(File.Exists(fileOptionStr)) { // File
					retVal = ReadTextFile(fileOptionStr);
				}
				else { // Entry Id or raw sequence
					retVal = fileOptionStr;
				}
			}
			PrintDebugMessage("LoadData", "End", 1);
			return retVal;
		}
		
		/// <summary>
		/// Load binary data for submission to the tool service.
		/// </summary>
		/// <param name="fileOptionStr">Name of file to read data from, a file name of "-" reads data from standard input (STDIN), or raw data to be sent to tool.</param>
		/// <returns>Data as a byte array.</returns>
		protected byte[] LoadBinData(string fileOptionStr) {
			PrintDebugMessage("LoadBinData", "Begin", 1);
			PrintDebugMessage("LoadBinData", "fileOptionStr: " + fileOptionStr, 2);
			byte[] retVal = null;
			if(fileOptionStr != null) {
				if(fileOptionStr == "-") { // STDIN
					retVal = ReadFile(fileOptionStr);
				}
				else if(File.Exists(fileOptionStr)) { // File
					retVal = ReadFile(fileOptionStr);
				}
				else { // Entry Id or raw sequence
					System.Text.ASCIIEncoding enc = new System.Text.ASCIIEncoding();
					retVal = enc.GetBytes(fileOptionStr);
				}
			}
			PrintDebugMessage("LoadBinData", "End", 1);
			return retVal;
		}
		
		/// <summary>
		/// Write a byte array to a file.
		/// </summary>
		/// <param name="fileName">File to write data to.</param>
		/// <param name="content">Data to write to file.</param>
		protected void WriteBinaryFile(string fileName, byte[] content) {
			PrintDebugMessage("WriteBinaryFile", "Begin", 1);
			PrintDebugMessage("WriteBinaryFile", "fileName: " + fileName, 1);
			PrintDebugMessage("WriteBinaryFile", "content: " + content.Length + " bytes", 1);
			if(fileName == "-") { // STDOUT
				Stream s = Console.OpenStandardOutput();
				BinaryWriter sw = new BinaryWriter(s);
				sw.Write(content);
				// Do not close, since this is STDOUT.
			}
			else { // Data file
				File.WriteAllBytes(fileName, content);
				Console.WriteLine("Wrote: {0}", fileName);
			}
			PrintDebugMessage("WriteBinaryFile", "End", 1);
		}

		/// <summary>
		/// Write text data encoded as a byte array to a file.
		/// </summary>
		/// <param name="fileName">File to write data to.</param>
		/// <param name="content">Text data to write to file.</param>
		protected void WriteTextFile(string fileName, byte[] content)
		{
			PrintDebugMessage("WriteTextFile", "Begin", 1);
			PrintDebugMessage("WriteTextFile", "fileName: " + fileName, 1);
			PrintDebugMessage("WriteTextFile", "content: " + content.Length + " bytes", 1);
			System.Text.ASCIIEncoding enc = new System.Text.ASCIIEncoding();
			string contentStr = enc.GetString(content);
			WriteTextFile(fileName, contentStr);
			PrintDebugMessage("WriteTextFile", "End", 1);
		}

		/// <summary>
		/// Write a string to a file.
		/// </summary>
		/// <param name="fileName">File to write data to.</param>
		/// <param name="content">Data to write to file.</param>
		protected void WriteTextFile(string fileName, string content)
		{
			PrintDebugMessage("WriteTextFile", "Begin", 1);
			PrintDebugMessage("WriteTextFile", "fileName: " + fileName, 1);
			PrintDebugMessage("WriteTextFile", "content: " + content.Length + " characters", 1);
			if (fileName == "-")
			{ // STDOUT
				Console.Write(content);
			}
			else
			{ // Data file
				File.WriteAllText(fileName, content);
				Console.WriteLine("Wrote: {0}", fileName);
			}
			PrintDebugMessage("WriteTextFile", "End", 1);
		}

		/// <summary>
		/// Get the service connection. Has to be called before attempting to use any of the service operations.
		/// </summary>
		protected abstract void ServiceProxyConnect();
		
		/// <summary>
		/// Get list of input parameter names from sevice.
		/// </summary>
		/// <returns>An array of parameter names.
		/// A <see cref="System.String"/>
		/// </returns>
		public abstract string[] GetParams();
		
		/// <summary>
		/// Print a list of input parameter names.
		/// </summary>
		protected void PrintParams() {
			PrintDebugMessage("PrintParams", "Begin", 1);
			string[] paramNameList = GetParams();
			foreach(string paramName in paramNameList) Console.WriteLine(paramName);
			PrintDebugMessage("PrintParams", "End", 1);
		}
		
		// Get details for a prameter.
		
		/// <summary>
		/// Print a detailed description of a specified input parameter.
		/// </summary>
		/// <param name="paramName">Name of the parameter to print the detailed description of.
		/// A <see cref="System.String"/>
		/// </param>
		protected abstract void PrintParamDetail(string paramName);
		
		/// <summary>
		/// Submit a job using the current client state.
		/// </summary>
		public abstract void SubmitJob();
		
		/// <summary>
		/// Get the status of a submitted job.
		/// </summary>
		/// <param name="jobId">Job identifier.
		/// A <see cref="System.String"/>
		/// </param>
		/// <returns>A string describing the status of the job.
		/// A <see cref="System.String"/>
		/// </returns>
		public abstract string GetStatus(string jobId);
		
		/// <summary>
		/// Print the status of the current job.
		/// </summary>
		public void PrintStatus() {
			PrintDebugMessage("PrintStatus", "Begin", 1);
			string status = GetStatus(JobId);
			Console.WriteLine(status);
			PrintDebugMessage("PrintStatus", "End", 1);
		}
		
		/// <summary>
		/// Wait for a job to finish.
		/// </summary>
		/// <param name="jobId">Job identifier of the job to wait for.
		/// A <see cref="System.String"/>
		/// </param>
		public void ClientPoll(string jobId) {
			PrintDebugMessage("ClientPoll", "Begin", 1);
			PrintDebugMessage("ClientPoll", "jobId: " + jobId, 2);
			int checkInterval = 1000;
			string status = "PENDING";
			// Check status and wait if not finished
			while(status == "RUNNING" || status == "PENDING") {
				status = GetStatus(JobId);
				PrintProgressMessage(status, 1);
				if(status == "RUNNING" || status == "PENDING") {
					// Wait before polling again.
					PrintDebugMessage("clientPoll", "checkInterval: " + checkInterval, 2);
					System.Threading.Thread.Sleep(checkInterval);
					checkInterval *= 2;
					if(checkInterval > MaxCheckInterval) checkInterval = MaxCheckInterval;

				}
			}
			PrintDebugMessage("ClientPoll", "End", 1);
		}
		
		/// <summary>
		/// Print a list of the available result types for the current job.
		/// </summary>
		public abstract void PrintResultTypes();
		
		/// <summary>
		/// Get results for a job, of the specified type.
		/// </summary>
		/// <param name="jobId">Job identifier.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="format">Required result format name.
		/// A <see cref="System.String"/>
		/// </param>
		public abstract byte[] GetResult(string jobId, string format);
		
		/// <summary>
		/// Get results for a job, of the specified type.
		/// </summary>
		/// <param name="jobId">Job identifier.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="outformat">Required result format. If null all available formats are retrived.
		/// A <see cref="System.String"/>
		/// </param>
		/// <param name="outFileBase">Base of generated filenames, used to store results. If null the filename is based on the job identifier.
		/// A <see cref="System.String"/>
		/// </param>
		public abstract void GetResults(string jobId, string outformat, string outFileBase);
		
		/// <summary>
		/// Get results for a job using the current format and output file.
		/// </summary>
		/// <param name="jobId">Job identifier.
		/// A <see cref="System.String"/>
		/// </param>
		public void GetResults(string jobId) {
			PrintDebugMessage("PollJob", "Begin", 1);
			GetResults(jobId, OutFormat, OutFile);
			PrintDebugMessage("PollJob", "End", 1);
		}
		
		/// <summary>
		/// Get results for the current job
		/// </summary>
		public void GetResults() {
			PrintDebugMessage("PollJob", "Begin", 1);
			GetResults(JobId, OutFormat, OutFile);
			PrintDebugMessage("PollJob", "End", 1);
		}
	}
}
